File: javaexample/src/main/java/io/chthonic/mythos/javaexample/ui/presenters/BasePresenter.java
package io.chthonic.mythos.javaexample.ui.presenters;

import android.os.Bundle;
import androidx.annotation.NonNull;

import io.chthonic.mythos.mvp.Presenter;
import io.chthonic.mythos.mvp.Vu;

/**
 * Created by jhavatar on 8/12/2018.
 */
public abstract class BasePresenter<V extends Vu> extends Presenter<V>  {
    private static final String SAVE_KEY = "save_key";

    private int saveCount = 0;

    protected String getText(String name) {
        return name + ": onSaveState count = " + saveCount + ", firstLink = " + getFirstLink() + ", firstLinkWithVu = " + getFirstLinkWithVu();
    }

    @Override
    public void onLink(@NonNull V vu, Bundle inState, @NonNull Bundle args) {
        super.onLink(vu, inState, args);
        if (inState != null) {
            saveCount = inState.getInt(SAVE_KEY, saveCount);
        }
    }

    @Override
    public void onSaveState(@NonNull Bundle outState) {
        super.onSaveState(outState);
        saveCount++;
        outState.putInt(SAVE_KEY, saveCount);
    }
}

File: kotlinexample/src/main/java/io/chthonic/mythos/kotlinexample/ui/presenters/BasePresenter.kt
package io.chthonic.mythos.kotlinexample.ui.presenters

import android.os.Bundle
import io.chthonic.mythos.mvp.Presenter
import io.chthonic.mythos.mvp.Vu

/**
 * Created by jhavatar on 8/12/2018.
 */
abstract class BasePresenter<V>: Presenter<V>() where V : Vu {
    companion object {
        private const val SAVE_KEY = "save_key"
    }
    private var saveCount = 0

    protected fun getText(name: String): String {
        return "$name: onSaveState count = $saveCount, firstLink = $firstLink, firstLinkWithVu = $firstLinkWithVu"
    }

    override fun onLink(vu: V, inState: Bundle?, args: Bundle) {
        super.onLink(vu, inState, args)
        saveCount = inState?.getInt(SAVE_KEY, saveCount) ?: saveCount
    }

    override fun onSaveState(outState: Bundle) {
        super.onSaveState(outState)
        saveCount++
        outState.putInt(SAVE_KEY, saveCount)
    }
}